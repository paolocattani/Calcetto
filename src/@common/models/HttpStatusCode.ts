// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
export enum HTTPStatusCode {
	Continue = 100,
	SwitchingProtocols = 101,
	Processing = 102,
	EarlyHints = 103,

	/**
	 * All `1xx` status codes.
	 */
	InformationalResponses = Continue | SwitchingProtocols | Processing | EarlyHints,

	OK = 200,
	Created = 201,
	Accepted = 202,
	NonAuthoritativeInformation = 203,
	NoContent = 204,
	ResetContent = 205,
	PartialContent = 206,
	MultiStatus = 207,
	AlreadyReported = 208,
	IMUsed = 226,

	/**
	 * All `2xx` status codes.
	 */
	Success = OK |
		Created |
		Accepted |
		NonAuthoritativeInformation |
		NoContent |
		ResetContent |
		PartialContent |
		MultiStatus |
		AlreadyReported |
		IMUsed,

	MultipleChoices = 300,
	MovedPermanently = 301,
	Found = 302,
	SeeOther = 303,
	NotModified = 304,
	UseProxy = 305,
	SwitchProxy = 306,
	TemporaryRedirect = 307,
	PermanentRedirect = 308,

	/**
	 * All `3xx` status codes.
	 */
	Redirection = MultipleChoices |
		MovedPermanently |
		Found |
		SeeOther |
		NotModified |
		UseProxy |
		SwitchProxy |
		TemporaryRedirect |
		PermanentRedirect,

	BadRequest = 400,
	/**
	 * The HTTP 401 Unauthorized client error status response code indicates that the request has not been applied because it lacks valid authentication credentials for the target resource.
	 */
	Unauthorized = 401,
	PaymentRequired = 402,
	/**
	 * The HTTP 403 Forbidden client error status response code indicates that the server understood the request but refuses to authorize it.
	 * This status is similar to 401, but in this case, re-authenticating will make no difference.
	 * The access is permanently forbidden and tied to the application logic, such as insufficient rights to a resource.
	 */
	Forbidden = 403,
	NotFound = 404,
	MethodNotAllowed = 405,
	NotAcceptable = 406,
	ProxyAuthenticationRequired = 407,
	RequestTimeout = 408,
	Conflict = 409,
	Gone = 410,
	LengthRequired = 411,
	PreconditionFailed = 412,
	PayloadTooLarge = 413,
	URITooLong = 414,
	UnsupportedMediaType = 415,
	RangeNotSatisfiable = 416,
	ExpectationFailed = 417,
	ImATeapot = 418,
	MisdirectedRequest = 421,
	UnprocessableEntity = 422,
	Locked = 423,
	FailedDependency = 424,
	UpgradeRequired = 426,
	PreconditionRequired = 428,
	TooManyRequests = 429,
	RequestHeaderFieldsTooLarge = 431,
	UnavailableForLegalReasons = 451,

	/**
	 * All `4xx` error codes.
	 */
	ClientErrors = BadRequest |
		Unauthorized |
		PaymentRequired |
		Forbidden |
		NotFound |
		MethodNotAllowed |
		NotAcceptable |
		ProxyAuthenticationRequired |
		RequestTimeout |
		Conflict |
		Gone |
		LengthRequired |
		PreconditionFailed |
		PayloadTooLarge |
		URITooLong |
		UnsupportedMediaType |
		RangeNotSatisfiable |
		ExpectationFailed |
		ImATeapot |
		MisdirectedRequest |
		UnprocessableEntity |
		Locked |
		FailedDependency |
		UpgradeRequired |
		PreconditionRequired |
		TooManyRequests |
		RequestHeaderFieldsTooLarge |
		UnavailableForLegalReasons,

	InternalServerError = 500,
	NotImplemented = 501,
	BadGateway = 502,
	ServiceUnavailable = 503,
	GatewayTimeout = 504,
	HTTPVersionNotSupported = 505,
	VariantAlsoNegotiates = 506,
	InsufficientStorage = 507,
	LoopDetected = 508,
	NotExtended = 510,
	NetworkAuthenticationRequired = 511,

	/**
	 * All `5xx` error codes.
	 */
	ServerErrors = InternalServerError |
		NotImplemented |
		BadGateway |
		ServiceUnavailable |
		GatewayTimeout |
		HTTPVersionNotSupported |
		VariantAlsoNegotiates |
		InsufficientStorage |
		LoopDetected |
		NotExtended |
		NetworkAuthenticationRequired,
}

export const InformationCodes = [
	HTTPStatusCode.Continue,
	HTTPStatusCode.SwitchingProtocols,
	HTTPStatusCode.Processing,
	HTTPStatusCode.EarlyHints,
];
export const SuccessCodes = [
	HTTPStatusCode.OK,
	HTTPStatusCode.Created,
	HTTPStatusCode.Accepted,
	HTTPStatusCode.NonAuthoritativeInformation,
	HTTPStatusCode.NoContent,
	HTTPStatusCode.ResetContent,
	HTTPStatusCode.PartialContent,
	HTTPStatusCode.MultiStatus,
	HTTPStatusCode.AlreadyReported,
	HTTPStatusCode.IMUsed,
];

export const RedirectCodes = [
	HTTPStatusCode.MultipleChoices,
	HTTPStatusCode.MovedPermanently,
	HTTPStatusCode.Found,
	HTTPStatusCode.SeeOther,
	HTTPStatusCode.NotModified,
	HTTPStatusCode.UseProxy,
	HTTPStatusCode.SwitchProxy,
	HTTPStatusCode.TemporaryRedirect,
	HTTPStatusCode.PermanentRedirect,
];

export const ClientErrorsCodes = [
	HTTPStatusCode.BadRequest,
	HTTPStatusCode.Unauthorized,
	HTTPStatusCode.PaymentRequired,
	HTTPStatusCode.Forbidden,
	HTTPStatusCode.NotFound,
	HTTPStatusCode.MethodNotAllowed,
	HTTPStatusCode.NotAcceptable,
	HTTPStatusCode.ProxyAuthenticationRequired,
	HTTPStatusCode.RequestTimeout,
	HTTPStatusCode.Conflict,
	HTTPStatusCode.Gone,
	HTTPStatusCode.LengthRequired,
	HTTPStatusCode.PreconditionFailed,
	HTTPStatusCode.PayloadTooLarge,
	HTTPStatusCode.URITooLong,
	HTTPStatusCode.UnsupportedMediaType,
	HTTPStatusCode.RangeNotSatisfiable,
	HTTPStatusCode.ExpectationFailed,
	HTTPStatusCode.ImATeapot,
	HTTPStatusCode.MisdirectedRequest,
	HTTPStatusCode.UnprocessableEntity,
	HTTPStatusCode.Locked,
	HTTPStatusCode.FailedDependency,
	HTTPStatusCode.UpgradeRequired,
	HTTPStatusCode.PreconditionRequired,
	HTTPStatusCode.TooManyRequests,
	HTTPStatusCode.RequestHeaderFieldsTooLarge,
	HTTPStatusCode.UnavailableForLegalReasons,
];

export const ServerErrorsCodes = [
	HTTPStatusCode.InternalServerError,
	HTTPStatusCode.NotImplemented,
	HTTPStatusCode.BadGateway,
	HTTPStatusCode.ServiceUnavailable,
	HTTPStatusCode.GatewayTimeout,
	HTTPStatusCode.HTTPVersionNotSupported,
	HTTPStatusCode.VariantAlsoNegotiates,
	HTTPStatusCode.InsufficientStorage,
	HTTPStatusCode.LoopDetected,
	HTTPStatusCode.NotExtended,
	HTTPStatusCode.NetworkAuthenticationRequired,
];
