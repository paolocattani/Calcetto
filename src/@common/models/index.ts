export * from './root.model';
export * from './pair.model';
export * from './player.model';
export * from './tournament.model';
export * from './auth.model';
export * from './stage1.model';
export * from './stage2.model';
export * from './common.models';
export * from './stats.model';
