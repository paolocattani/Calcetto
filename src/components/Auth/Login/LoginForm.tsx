import * as React from 'react';
import { Form, Field, ErrorMessage } from 'formik';
import { Row, Col, FormControl, Button } from 'react-bootstrap';

const LoginForm: React.FC = () => (
  <Form>
    <Row>
      <Col>
        <label htmlFor="email">Email</label>
        <Field as={FormControl} id="email" name="email" placeholder="Email" type="email" />
        <ErrorMessage name="email" />
      </Col>
    </Row>

    <Row>
      <Col>
        <label htmlFor="password">Password</label>
        <Field as={FormControl} id="password" name="password" placeholder="Password" type="password" />
        <ErrorMessage name="password" />
      </Col>
    </Row>
    <Row>
      <Col>
        <Button type="submit">Login</Button>
      </Col>
    </Row>
  </Form>
);

export default LoginForm;
