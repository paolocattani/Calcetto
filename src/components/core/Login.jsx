// React
import React, { Component } from 'react';
import { Form, Button, Alert } from 'react-bootstrap';
// Router
import { withRouter } from 'react-router';

const formStyle = { width: 'fit-content', margin: 'auto' };

class LoginClass extends Component {
   constructor(props) {
      super(props);
      this.state = { username: '', password: '' };
      this.handleChange = this.handleChange.bind(this);
      this.handleSubmit = this.handleSubmit.bind(this);
   }

   componentDidMount() {
      console.log('Login.componentDidMount : props -> ', this.props);
   }

   handleChange(event) {
      this.setState({ [event.currentTarget.name]: event.currentTarget.value });
   }

   // chiamare API autenticazione
   async handleSubmit(event) {
      event.preventDefault();
      // Test Auth
      const { username, password } = this.state;
      console.log('Login.handleSubmit.userId', username);
      console.log('Login.handleSubmit.password', password);

      const response = await fetch('/api/alfresco/auth', {
         method: 'POST',
         headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
         body: JSON.stringify({ username, password })
      });
      const result = await response.json();
      const { handleLogin } = this.props;
      handleLogin(result.userId, result.id);
      // https://reacttraining.com/react-router/web/example/auth-workflow
      const { history } = this.props;
      const { location } = this.props;
      const { from } = location.state || { from: { pathname: '/' } };
      history.replace(from);
      // this.props.history.push("/");
   }

   // Render form username/password
   render() {
      const message = this.props.location?.state?.message ?? 'Effettua il login per continuare';
      return (
         <>
            {message ? (
               <Alert key={1} variant="danger">
                  {message}
               </Alert>
            ) : null}
            <Form style={formStyle} onSubmit={this.handleSubmit}>
               <Form.Group controlId="username">
                  <Form.Label>Username</Form.Label>
                  <Form.Control name="username" type="text" placeholder="username" onChange={this.handleChange} />
               </Form.Group>

               <Form.Group controlId="password">
                  <Form.Label>Password</Form.Label>
                  <Form.Control name="password" type="password" placeholder="Password" onChange={this.handleChange} />
               </Form.Group>
               <Button variant="warning" type="submit">
                  {' '}
                  Login{' '}
               </Button>
            </Form>
         </>
      );
   }
}

const Login = withRouter(LoginClass);
export default Login;
