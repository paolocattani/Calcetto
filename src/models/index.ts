export * from './common.model';
export * from './pair.model';
export * from './player.model';
export * from './tournament.model';
export * from './session.model';
export * from './stage1.model';
export * from './stage2.model';
