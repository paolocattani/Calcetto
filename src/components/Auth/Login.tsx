import { useInput } from '../core/hooks/InputHook';
import { Form, Button, Row, Col } from 'react-bootstrap';
import React from 'react';
import { useHistory, RouteComponentProps, withRouter } from 'react-router-dom';
import { SessionAction, TournamentAction } from 'redux/actions';
import { useDispatch } from 'react-redux';
import { UserDTO } from 'redux/models/user.model';
import { UserMessage } from 'redux/models';
import { toast } from 'react-toastify';

interface PropsType extends RouteComponentProps {}

// https://medium.com/@faizanv/authentication-for-your-react-and-express-application-w-json-web-tokens-923515826e0#6563
const Login: React.FC<PropsType> = (): JSX.Element => {
  const dispatch = useDispatch();
  const currentHistory = useHistory();

  const { value: username, bind: bindUsername } = useInput<string>('');
  const { value: password, bind: bindPassword } = useInput<string>('');

  const handleSubmit = async (evt: React.SyntheticEvent) => {
    evt.preventDefault();
    if (!username || username === '') {
      toast.error('Inserire username o password');
      return;
    }
    if (!password || password === '') {
      toast.error('Inserire la password');
      return;
    }
    try {
      const response = await fetch('/api/v1/auth/authenticate', {
        method: 'POST',
        body: JSON.stringify({ username, password }),
        headers: { 'Content-Type': 'application/json' },
      });
      const result: UserDTO = await response.json();
      if (response.ok && result) {
        // Messaggio di benvenuto
        const message: UserMessage = { type: 'success', message: `Bentornato ${result.username} !` };
        toast.success(message.message);
        dispatch(SessionAction.updateSession({ user: result, message, showMessage: true }));
        dispatch(TournamentAction.fetchTournaments.request({}));
        dispatch(
          SessionAction.sessionControl.request({
            history: currentHistory,
          })
        );
        currentHistory.push('/');
      } else {
        if (response.status === 401) toast.error('Utente o Password errata');
        else toast.error('Errore server. Riprovare piu tardi');
      }
    } catch (error) {
      console.log('onSubmitLogin : ', error);
      toast.error('Errore durante il processo di registrazione. Riprovare piu tardi');
    }
  };

  return (
    <Form onSubmit={handleSubmit}>
      <Form.Group controlId="username">
        <Form.Label>Username o Email</Form.Label>
        <Form.Control required type="text" placeholder="username o email" {...bindUsername} />
      </Form.Group>

      <Form.Group controlId="password">
        <Form.Label>Password</Form.Label>
        <Form.Control required type="password" placeholder="Password" {...bindPassword} />
      </Form.Group>
      <Row>
        <Col>
          <Button size="lg" variant="outline-success" className="float-right" type="submit">
            Conferma
          </Button>
        </Col>
      </Row>
    </Form>
  );
};

export default withRouter(Login);
