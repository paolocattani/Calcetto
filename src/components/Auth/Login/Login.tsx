import * as React from 'react';

import { Formik } from 'formik';
import * as yup from 'yup';
import LoginForm from './LoginForm';
import { ILoginFormValue, ILoginForm } from '../types';

const initialValues: ILoginFormValue = {
  email: '',
  password: ''
};

const validationSchema = yup.object().shape({ password: yup.string(), email: yup.string() });

const Login: React.FC<ILoginForm> = ({ onSubmit }: ILoginForm) => (
  <Formik validationSchema={validationSchema} initialValues={initialValues} onSubmit={onSubmit}>
    <LoginForm />
  </Formik>
);

export default Login;
