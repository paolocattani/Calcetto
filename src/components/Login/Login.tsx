import React from 'react';
import { useSessionContext } from '../core/SessionContext';
import { useHistory } from 'react-router';

// TODO:
export const Login: React.FC = () => {
  const [sessionContext, updateSessionContext] = useSessionContext();
  const currentHistory = useHistory();

  const handleClick = () => {
    updateSessionContext({ ...sessionContext, isAuthenticated: true });
    currentHistory.push('/');
  };

  return <button onClick={handleClick}>Login</button>;
};
