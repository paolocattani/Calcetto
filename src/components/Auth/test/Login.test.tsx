import React from 'react';
import Login from '../Login';
import { render, RenderResult, fireEvent, screen } from '../../../test/test-utils';
import userEvent from '@testing-library/user-event';
import loginResponse from './_mocks_/login_response.json';
import { HTTPStatusCode } from 'src/@common/models/HttpStatusCode';
import fetchMock from 'jest-fetch-mock';

describe('<Login />.render', () => {
	let component: RenderResult;
	beforeEach(async () => {
		component = render(<Login />);
	});

	it('should render correctly should match snapshot', () => {
		expect(component).toMatchSnapshot();
	});
	it('should contains all elements', async () => {
		await screen.findByRole('button', { name: /confirm/ });
		expect(screen.getByLabelText(/username/i)).not.toBeNull();
		expect(screen.getByLabelText(/password/i)).not.toBeNull();
		expect(screen.getByRole('button', { name: 'confirm' })).not.toBeNull();
	});
	it('input elements should have "required" attritbute', async () => {
		await screen.findByRole('button', { name: /confirm/i });
		expect(screen.getByLabelText(/username/i)).toBeRequired();
		expect(screen.getByLabelText(/password/i)).toBeRequired();
	});
});

//https://kentcdodds.com/blog/avoid-nesting-when-youre-testing?fbclid=IwAR2snxztWxnzUQGzIG42DrmqedSrzmmdfYgBWva7_Vzb3bgj6QR8UazP5_4
// https://www.robinwieruch.de/react-testing-library
// https://kentcdodds.com/blog/common-mistakes-with-react-testing-library

// GREAT : https://jkettmann.com/beginners-guide-to-testing-react/
describe('<Login /> tests ', () => {
	let usernameField: HTMLElement;
	let passwordField: HTMLElement;
	let loginButton: HTMLElement;
	let changeUsernameInput: (value: string) => boolean;
	let changePasswordInput: (value: string) => boolean;
	const user = { username: 'michelle', password: 'smith' };

	beforeEach(async () => {
		render(<Login />);
		usernameField = screen.getByLabelText(/username/i);
		passwordField = screen.getByLabelText(/password/i);
		loginButton = await screen.findByRole('button', { name: /confirm/i });
		changeUsernameInput = (value: string) => fireEvent.change(usernameField, { target: { value } });
		changePasswordInput = (value: string) => fireEvent.change(passwordField, { target: { value } });
	});

	describe('when username and password is not provided', () => {
		it('should have empty fields and button disabled ', async () => {
			expect(usernameField.getAttribute('value')).toEqual('');
			expect(passwordField.getAttribute('value')).toEqual('');
			expect(loginButton).toBeDisabled();
		});
	});

	describe('when username and password is provided', () => {
		beforeEach(() => {
			changeUsernameInput(user.username);
			changePasswordInput(user.password);
		});

		it('should have changed field value', () => {
			expect(usernameField.getAttribute('value')).toEqual(user.username);
			expect(passwordField.getAttribute('value')).toEqual(user.password);
			expect(loginButton).not.toBeDisabled();
		});

		it('should type value', () => {
			userEvent.clear(usernameField);
			userEvent.clear(passwordField);
			userEvent.type(usernameField, user.username);
			userEvent.type(passwordField, user.password);
			expect(usernameField.getAttribute('value')).toEqual(user.username);
			expect(passwordField.getAttribute('value')).toEqual(user.password);
		});

		/* FIXME:
    describe('when the submit button is clicked', () => {
      beforeEach(() => {
        fetchMock.resetMocks();
      });
      it('should call onSubmit with the username and password', () => {
        userEvent.click(loginButton);
        fetchMock.once('http://localhost:5001/api/v2/auth/login', { status: HTTPStatusCode.OK });

        expect(fetch).toHaveBeenCalledTimes(1);
        expect(fetch).toHaveBeenCalledWith('/api/v2/auth/login', {
          method: 'POST',
          body: JSON.stringify({ username: user.username, password: user.password }),
          headers: { 'Content-Type': 'application/json' },
        });
      });
    });
    */
	});
});
