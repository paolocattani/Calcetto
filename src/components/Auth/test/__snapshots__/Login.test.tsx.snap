// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<Login />.render should render correctly should match snapshot 1'] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="fade modal-backdrop show"
    />
    <div
      aria-modal="true"
      class="fade modal show"
      role="dialog"
      style="border-color: #ffc107; border-width: 3px; display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered"
        data-cy="loader"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
            style="color: whitesmoke; background-color: rgb(52, 58, 64); border-color: #ffc107;"
          >
            <div
              class="modal-title h4"
            >
              Caricamento....
            </div>
            <button
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
              <span
                class="sr-only"
              >
                Close
              </span>
            </button>
          </div>
          <div
            class="text-center font-weight-bold modal-body"
            style="color: whitesmoke; background-color: rgb(52, 58, 64); border-color: #ffc107;"
          >
            Caricamento....
          </div>
          <div
            class="modal-footer"
            style="color: whitesmoke; background-color: rgb(52, 58, 64); border-color: #ffc107;"
          >
            <div
              class="spinner-border text-light"
            />
            <div
              class="spinner-border text-primary"
            />
            <div
              class="spinner-border text-secondary"
            />
            <div
              class="spinner-border text-success"
            />
            <div
              class="spinner-border text-danger"
            />
            <div
              class="spinner-border text-warning"
            />
            <div
              class="spinner-border text-info"
            />
            <div
              class="spinner-border text-dark"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
